core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3963,6 +3969,60 @@ private void handleSuccess(PersistJob job) {
+                // check if the destination direction is valid, if there isn't exist directory,
+                // create it and it's parents
+                Stack<Pair<String, Inode>> ancestors = new Stack<>();
+                List<Inode> inodes = inodePath.getInodeList();
+                int curInodeIndex = inodes.size() - 2;
+                // get file path
+                AlluxioURI curUfsPath = new AlluxioURI(ufsPath);
+                // get the parent path of current file
+                curUfsPath = curUfsPath.getParent();
+                // Stop when the directory already exists in UFS.
+                while (!ufs.isDirectory(curUfsPath.toString()) && curInodeIndex >= 0) {
+                  Inode curInode = inodes.get(curInodeIndex);
+                  ancestors.push(new Pair<>(curUfsPath.toString(), curInode));
+                  curUfsPath = curUfsPath.getParent();
+                  curInodeIndex--;
+                }
+
+                while (!ancestors.empty()) {
+                  Pair<String, Inode> ancestor = ancestors.pop();
+                  String dir = ancestor.getFirst();
+                  Inode ancestorInode = ancestor.getSecond();
+                  MkdirsOptions options = MkdirsOptions.defaults(ServerConfiguration.global())
+                      .setCreateParent(false)
+                      .setOwner(ancestorInode.getOwner())
+                      .setGroup(ancestorInode.getGroup())
+                      .setMode(new Mode(ancestorInode.getMode()));
+                  // UFS mkdirs might fail if the directory is already created.
+                  // If so, skip the mkdirs and assume the directory is already prepared,
+                  // regardless of permission matching.
+                  try {
+                    boolean mkdirSuccess = false;
+                    try {
+                      mkdirSuccess = ufs.mkdirs(dir, options);
+                    } catch (IOException e) {
+                      LOG.debug("Exception Directory {}: ", dir, e);
+                    }
+                    if (mkdirSuccess) {
+                      LOG.debug("UFS directory {} is created", dir);
+                      List<AclEntry> allAcls =
+                          Stream.concat(ancestorInode.getDefaultACL().getEntries().stream(),
+                          ancestorInode.getACL().getEntries().stream())
+                              .collect(Collectors.toList());
+                      ufs.setAclEntries(dir, allAcls);
+                    } else {
+                      if (ufs.isDirectory(dir)) {
+                        LOG.debug("UFS directory {} already exists", dir);
+                      } else {
+                        LOG.info("UFS path {} is an existing file", dir);
+                      }
+                    }
+                  } catch (IOException e) {
+                    LOG.error("Failed to create UFS directory {} with correct permission", dir, e);
+                  }
+                }
```

Comment:
Should this logic be a method?

Reply:
Yeah, it make sense, changes to `createParentPath` method.